File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/signal/STFT.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  *  See the NOTICE file distributed with this work for additional
 *  *  information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.signal;

import lombok.NoArgsConstructor;
import lombok.NonNull;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.common.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

/**
 * Short-Time Fourier Transform operation.
 * Computes STFT by applying DFT to windowed overlapping segments.
 *
 * @author Adam Gibson
 */
@NoArgsConstructor
public class STFT extends DynamicCustomOp {
    private boolean onesided = true;

    public STFT(@NonNull SameDiff sameDiff, @NonNull SDVariable signal, @NonNull SDVariable frameStep,
                SDVariable window, SDVariable frameLength, boolean onesided) {
        super(sameDiff, buildInputs(signal, frameStep, window, frameLength));
        this.onesided = onesided;
        addArgs();
    }

    public STFT(@NonNull SameDiff sameDiff, @NonNull SDVariable signal, @NonNull SDVariable frameStep) {
        this(sameDiff, signal, frameStep, null, null, true);
    }

    public STFT(@NonNull SameDiff sameDiff, @NonNull SDVariable signal, @NonNull SDVariable frameStep, boolean onesided) {
        this(sameDiff, signal, frameStep, null, null, onesided);
    }

    public STFT(@NonNull INDArray signal, @NonNull INDArray frameStep, INDArray window, INDArray frameLength,
                boolean onesided, INDArray output) {
        super(buildInputs(signal, frameStep, window, frameLength), wrapOrNull(output));
        this.onesided = onesided;
        addArgs();
    }

    public STFT(@NonNull INDArray signal, @NonNull INDArray frameStep, INDArray window, INDArray frameLength,
                boolean onesided) {
        this(signal, frameStep, window, frameLength, onesided, null);
    }

    /**
     * Simplified STFT constructor without window/frameLength for codegen.
     */
    public STFT(@NonNull INDArray signal, @NonNull INDArray frameStep, boolean onesided) {
        this(signal, frameStep, null, null, onesided, null);
    }

    private static SDVariable[] buildInputs(SDVariable signal, SDVariable frameStep, SDVariable window, SDVariable frameLength) {
        if (window != null && frameLength != null) {
            return new SDVariable[]{signal, frameStep, window, frameLength};
        } else if (window != null) {
            return new SDVariable[]{signal, frameStep, window};
        } else {
            return new SDVariable[]{signal, frameStep};
        }
    }

    private static INDArray[] buildInputs(INDArray signal, INDArray frameStep, INDArray window, INDArray frameLength) {
        if (window != null && frameLength != null) {
            return new INDArray[]{signal, frameStep, window, frameLength};
        } else if (window != null) {
            return new INDArray[]{signal, frameStep, window};
        } else {
            return new INDArray[]{signal, frameStep};
        }
    }

    @Override
    public String opName() {
        return "stft";
    }

    protected void addArgs() {
        addIArgument(onesided ? 1 : 0);
    }

    @Override
    public List<SDVariable> doDiff(List<SDVariable> grad) {
        throw new UnsupportedOperationException("Gradient for STFT not implemented");
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes) {
        Preconditions.checkState(inputDataTypes != null && inputDataTypes.size() >= 2,
                "Expected at least 2 input datatypes for %s, got %s", getClass(), inputDataTypes);
        return Collections.singletonList(DataType.FLOAT);
    }
}

File: libnd4j/include/ops/declarable/generic/signal/stft.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author Adam Gibson
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_stft)

#include <ops/declarable/headers/signal.h>
#include <ops/declarable/helpers/signal.h>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(stft, 2, 1, false, 0, -2) {
    auto signal = INPUT_VARIABLE(0);     // [batch, signal_length] or [batch, signal_length, 1]
    auto frameStep = INPUT_VARIABLE(1);  // scalar
    auto output = OUTPUT_VARIABLE(0);

    NDArray* window = nullptr;
    int frameLength = 256;  // Default frame length

    if (block.width() > 2) {
        window = INPUT_VARIABLE(2);
    }
    if (block.width() > 3) {
        auto frameLengthTensor = INPUT_VARIABLE(3);
        frameLength = frameLengthTensor->e<int>(0);
    } else if (window != nullptr) {
        frameLength = static_cast<int>(window->lengthOf());
    }

    int frameStepVal = frameStep->e<int>(0);
    bool onesided = block.numI() > 0 ? INT_ARG(0) != 0 : true;

    REQUIRE_TRUE(signal->rankOf() >= 1 && signal->rankOf() <= 3, 0,
                 "stft: signal must be rank 1-3, got rank %i", signal->rankOf());

    helpers::stft(block.launchContext(), signal, frameStepVal, window, frameLength, onesided, output);

    return sd::Status::OK;
}

DECLARE_SHAPE_FN(stft) {
    auto signalShape = inputShape->at(0);
    auto frameStep = INPUT_VARIABLE(1);

    NDArray* window = nullptr;
    int frameLength = 256;

    if (block.width() > 2) {
        window = INPUT_VARIABLE(2);
    }
    if (block.width() > 3) {
        auto frameLengthTensor = INPUT_VARIABLE(3);
        frameLength = frameLengthTensor->e<int>(0);
    } else if (window != nullptr) {
        frameLength = static_cast<int>(window->lengthOf());
    }

    int frameStepVal = frameStep->e<int>(0);
    bool onesided = block.numI() > 0 ? INT_ARG(0) != 0 : true;

    auto signalRank = shape::rank(signalShape);
    sd::LongType batchSize = signalRank > 1 ? shape::sizeAt(signalShape, static_cast<sd::LongType>(0)) : 1;
    sd::LongType signalLength = signalRank > 1 ?
        shape::sizeAt(signalShape, static_cast<sd::LongType>(1)) :
        shape::sizeAt(signalShape, static_cast<sd::LongType>(0));

    sd::LongType numFrames = (signalLength - frameLength) / frameStepVal + 1;
    sd::LongType numBins = onesided ? frameLength / 2 + 1 : frameLength;

    // Output: [batch, num_frames, num_bins, 2]
    std::vector<sd::LongType> outputShapeVec = {batchSize, numFrames, numBins, 2};

    return SHAPELIST(ConstantShapeHelper::getInstance().createShapeInfo(sd::DataType::FLOAT32, 'c', outputShapeVec));
}

DECLARE_TYPES(stft) {
    getOpDescriptor()->addTraits(OP_TRAIT_FULLY_WRITING);
    getOpDescriptor()
        ->setAllowedInputTypes(0, {ALL_FLOATS})
        ->setAllowedInputTypes(1, {ALL_INTS})
        ->setAllowedInputTypes(2, {ALL_FLOATS})  // window (optional)
        ->setAllowedInputTypes(3, {ALL_INTS})    // frame_length (optional)
        ->setAllowedOutputTypes({ALL_FLOATS});
}

}  // namespace ops
}  // namespace sd

#endif
